osd: simply osd scenario declaration

leseb opened the pull request:
In order to choose a scenario, you have to use the osd_scenario variable
and set it appropriately, e.g: osd_scenario: raw_multi_journal

Signed-off-by: Sébastien Han <seb@redhat.com>

leseb commented:
I know I need to update the tests but for now, look at the structure only.

andrewschoen commented on roles/ceph-osd/tasks/check_devices.yml:
```diff
@@ -40,7 +40,7 @@
   register: journal_partition_status
   when:
-    - (raw_multi_journal or dmcrypt_dedicated_journal)
+    - (osd_scenario == 'raw_multi_journal' or osd_scenario == 'dmcrypt_dedicated_journal')
```
does ``(osd_scenario in ['raw_multi_journal', 'dmcrypt_dedicated_journal'])`` work here?

leseb replied:
done

andrewschoen commented on roles/ceph-osd/tasks/main.yml:
```diff
@@ -8,21 +8,21 @@
 - include: ./scenarios/journal_collocation.yml
   when:
-    - (journal_collocation or dmcrypt_journal_collocation)
+    - (osd_scenario == 'journal_collocation' or osd_scenario == 'dmcrypt_journal_collocation')
```
this is a good place for ``osd_scenario in ['journal_collocation', 'dmcrypt_journal_collocation']``

andrewschoen commented on roles/ceph-osd/tasks/check_mandatory_vars.yml:
```diff
@@ -26,32 +26,24 @@
-      or (dmcrypt_dedicated_journal and journal_collocation)
-      or (dmcrypt_dedicated_journal and raw_multi_journal)
-      or (dmcrypt_dedicated_journal and dmcrypt_journal_collocation)
+    - not osd_scenario in ['journal_collocation', 'raw_multi_journal', 'dmcrypt_dedicated_journal', 'dmcrypt_journal_collocation']
```
I'm thinking we should keep this list in ``defaults/main.yml``. Then we could also print the list as part of the failure message here.

leseb commented:
jenkins test jewel-ansible2.2-purge_cluster

leseb commented:
jenkins test luminous-ansible2.2-bluestore_cluster

leseb commented:
jenkins test jewel-ansible2.2-xenial_cluster

leseb commented:
jenkins test luminous-ansible2.2-centos7_cluster

leseb commented:
jenkins test luminous-ansible2.2-centos7_cluster

leseb commented:
jenkins test luminous-ansible2.2-journal_collocation

leseb commented:
jenkins test luminous-ansible2.2-purge_cluster

leseb commented:
jenkins test luminous-ansible2.2-purge_docker_cluster

leseb commented:
jenkins test luminous-ansible2.2-xenial_cluster

leseb commented:
jenkins test luminous-ansible2.2-update_docker_cluster

andrewschoen commented:
jenkins test luminous-ansible2.2-centos7_cluster

andrewschoen commented:
The failing ``*-purge_cluster`` and ``*-xenial_cluster`` tests are because this still needs to be updated: https://github.com/ceph/ceph-ansible/blob/simplify-osd-scenario/tests/functional/ubuntu/16.04/cluster/group_vars/all#L13

leseb commented:
jenkins test luminous-ansible2.2-centos7_cluster

leseb commented:
jenkins test luminous-ansible2.2-update_docker_cluster

leseb commented:
jenkins test luminous-ansible2.2-update_docker_cluster

leseb commented:
jenkins test jewel-ansible2.2-purge_cluster

andrewschoen commented on roles/ceph-osd/tasks/activate_osds.yml:
```diff
@@ -38,7 +38,7 @@
     - item.value.holders|count == 0
     - osd_auto_discovery
-    - dmcrypt_journal_collocation
+    - osd_scenario == 'dmcrypt_journal_collocation'
```
Shouldn't this be ``osd_scenario == 'journal_collocation'``? You'd need to check the ``dmcrypt`` variable as well.

andrewschoen commented on group_vars/osds.yml.sample:
```diff
@@ -115,22 +116,33 @@ dummy:
 # Use 'true' for 'raw_multi_journal' to enable this scenario
 # List devices under 'devices' variable above and
-# write journal devices for those under 'raw_journal_devices'
+# write journal devices for those under 'dedicated_devices'
```
How do you feel about renaming ``dedicated_devices`` to ``journal_devices``? I think ``journal`` needs to be in the name of the variable somehow. It's a bit confusing to have to define ``devices`` and ``dedicated_devices``, the difference is not very apparent.

andrewschoen commented on group_vars/osds.yml.sample:
```diff
@@ -115,22 +116,33 @@ dummy:
+#osd_scenario: dummy
+#valid_osd_scenarios:
+#  - collocated
+#  - non-collocated
```
So they are more self-explanatory should we use ``collocated_journals`` and ``dedicated_journals`` here instead?

andrewschoen commented on roles/ceph-osd/tasks/activate_osds.yml:
```diff
@@ -26,7 +26,7 @@
     - item.0.get("rc", 0) != 0
     - not osd_auto_discovery
-    - raw_multi_journal
+    - osd_scenario == 'raw_multi_journal'
```
This should check the new scenario name.

andrewschoen commented on roles/ceph-osd/tasks/activate_osds.yml:
```diff
@@ -52,13 +52,13 @@
     - item.0.get("rc", 0) != 0
     - not osd_auto_discovery
-    - dmcrypt_dedicated_journal
+    - osd_scenario == 'dmcrypt_dedicated_journal'
```
This needs to check the new scenario name and ``dmcrypt``.

andrewschoen commented on roles/ceph-osd/tasks/activate_osds.yml:
```diff
@@ -52,13 +52,13 @@
 - name: combine ispartition results
   set_fact:
-    combined_activate_osd_disk_results: "{{ activate_osd_disk if not dmcrypt_journal_collocation else activate_osd_disk_dmcrypt }}"
+    combined_activate_osd_disk_results: "{{ activate_osd_disk if osd_scenario != 'dmcrypt_journal_collocation' else activate_osd_disk_dmcrypt }}"
```
This needs to use the new scenario name as well.

andrewschoen commented on roles/ceph-osd/tasks/check_devices.yml:
```diff
@@ -34,22 +34,22 @@
   register: journal_partition_status
   when:
-    - (raw_multi_journal or dmcrypt_dedicated_journal)
+    - (osd_scenario in ['raw_multi_journal', 'dmcrypt_dedicated_journal'])
```
This also needs to check the new scenario names.

andrewschoen commented on roles/ceph-osd/tasks/pre_requisite.yml:
```diff
@@ -30,7 +30,7 @@
     copy_admin_key: true
   when:
-    - dmcrypt_journal_collocation or dmcrypt_dedicated_journal
+    - (osd_scenario in ['dmcrypt_journal_collocation', 'dmcrypt_dedicated_journal'])
```
This also needs to check the new scenario names.

andrewschoen commented:
I'm liking the direction this is going. It makes me wonder if we shouldn't eventually do something similar to the install options. Instead of having many variables to define how to install ceph, we could have a ``install_method`` variable or something that we can set like ``install_method: upstream``.

leseb commented:
@andrewschoen thanks for the review, I've addressed your comments. The only remaining thing I haven't changed (yet?) is the `dedicated_devices` variable name. I came up with this one using the same reasoning of `collocated` and `non-collocated`. Basically, I want this variable to be agnostic whether you use bluestore or filestore. For instance, in Bluestore, there is no such concept of 'journal' anymore, this is really a filestore thing. So if we use 'journal' I'm afraid users will be confused as this might look like filestore. I know the variable is not self-explanatory but at least it's agnostic.

What do you think?

leseb commented:
jenkins test jewel-ansible2.2-docker_cluster

leseb commented:
jenkins test jewel-ansible2.2-docker_dmcrypt_journal_collocation

leseb commented:
jenkins test luminous-ansible2.2-bluestore_journal_collocation

leseb commented:
jenkins test luminous-ansible2.2-docker_dmcrypt_journal_collocation

leseb commented:
jenkins test luminous-ansible2.2-bluestore_docker_dmcrypt_journal_collocation

leseb commented:
jenkins test pr-syntax-check

leseb commented:
jenkins test playbook syntax

leseb commented:
jenkins test luminous-ansible2.2-purge_dmcrypt

andrewschoen commented on tests/conftest.py:
```diff
@@ -31,7 +31,7 @@ def node(Ansible, Interface, Command, request):
         pytest.skip("mgr nodes can not be tested with ceph release jewel")
 
-    journal_collocation_test = ansible_vars.get("journal_collocation") or ansible_vars.get("dmcrypt_journal_collocation")
+    journal_collocation_test = ansible_vars.get("osd_scenario", 'collocated')
```
This needs to return a boolean. I think what you'll need here is: ``journal_collocation_test = ansible_vars.get("osd_scenario") == "collocated"``

leseb replied:
thanks!

andrewschoen commented on tox.ini:
```diff
@@ -148,4 +148,5 @@ commands=
   update_cluster: {[update]commands}
   update_docker_cluster: {[update]commands}
 
+  sleep 3h
```
Is this here for debugging purposes? This will slow down the CI by quite a lot.

leseb replied:
Yes, but this doesn't seem to be working...

andrewschoen replied:
I haven't looked at a log, but that's probably because ``sleep`` isn't whitelisted here https://github.com/ceph/ceph-ansible/blob/master/tox.ini#L56

ceph-jenkins commented:
FAIL - luminous-ansible2.2-dmcrypt_journal

leseb commented:
jenkins test this please

leseb commented:
jenkins test this please

andrewschoen commented on tests/functional/centos/7/docker-ded-jrn/group_vars/all:
```diff
@@ -14,9 +14,10 @@ public_network: "192.168.15.0/24"
 cluster_network: "192.168.16.0/24"
 ceph_rgw_civetweb_port: 8080
 ceph_osd_docker_devices: "{{ devices }}"
+osd_scenario: collocated
```
This needs to be ``osd_scenario: non-collocated``

leseb replied:
Good catch, thanks!

andrewschoen commented:
jenkins test jewel-ansible2.2-update_cluster

andrewschoen commented:
jenkins test luminous-ansible2.2-bluestore_docker_cluster

andrewschoen commented:
jenkins test luminous-ansible2.2-purge_docker_cluster

andrewschoen commented:
jenkins test luminous-ansible2.2-update_cluster

andrewschoen commented:
jenkins test luminous-ansible2.2-dmcrypt_journal_collocation

andrewschoen commented:
jenkins test luminous-ansible2.2-dmcrypt_journal_collocation

andrewschoen commented:
jenkins test luminous-ansible2.2-dmcrypt_journal

andrewschoen commented:
jenkins test jewel-ansible2.2-xenial_cluster

andrewschoen commented:
jenkins test jewel-ansible2.2-update_docker_cluster

andrewschoen commented:
jenkins test jewel-ansible2.2-journal_collocation

leseb commented:
jenkins test jewel-ansible2.2-dmcrypt_journal_collocation

leseb commented:
jenkins test jewel-ansible2.2-docker_cluster

leseb commented:
jenkins test luminous-ansible2.2-bluestore_journal_collocation

leseb commented:
jenkins test luminous-ansible2.2-purge_dmcrypt

leseb commented:
jenkins test luminous-ansible2.2-purge_dmcrypt

leseb commented:
jenkins test luminous-ansible2.2-bluestore_journal_collocation

leseb commented:
jenkins test luminous-ansible2.2-centos7_cluster

leseb commented:
jenkins test jewel-ansible2.2-update_cluster

leseb commented:
jenkins test jewel-ansible2.2-update_cluster

leseb commented:
jenkins test luminous-ansible2.2-bluestore_journal_collocation

leseb commented:
jenkins test luminous-ansible2.2-bluestore_journal_collocation

leseb commented:
jenkins test luminous-ansible2.2-bluestore_journal_collocation

leseb commented:
jenkins test luminous-ansible2.2-bluestore_journal_collocation

leseb commented:
jenkins test luminous-ansible2.2-bluestore_journal_collocation
